### PR TITLE
Move performance issues post_process to dedicated Celery queue

### PR DESCRIPTION
**Pull request by rjo100**
Sends dedicated issue category data to `post_process_group` call so we can route to the appropriate celery queue

### CODE REVIEW

File: src/sentry/eventstream/kafka/postprocessworker.py [hunk 1]
```diff
@@ -75,20 +76,24 @@ def dispatch_post_process_group_task(
...
+            kwargs={
+                "is_new": is_new,
+                "is_regression": is_regression,
+                "is_new_group_environment": is_new_group_environment,
+                "primary_hash": primary_hash,
+                "cache_key": cache_key,
+                "group_id": group_id,
+                "group_states": group_states,
+            },
+            queue=get_post_process_queue(group_category),
```
Comment (rjo100) [inline src/sentry/eventstream/kafka/postprocessworker.py:96]: don't totally understand the difference between this call and the other one. when does each get used?

Comment (wedamija) [reply src/sentry/eventstream/kafka/postprocessworker.py:96]: This call is used by the real Kafka post process worker, which is used in production. The other is used by `SnubaEventStream`, which just dispatches the task on eventstream insert instead.

Comment (barkbarkimashark) [reply src/sentry/eventstream/kafka/postprocessworker.py:96]: I believe this one isn't really used in prod. Not sure about the history but I can imagine this was maybe a reference implementation and the Snuba and/or Kafka ones came later and superceeded the base EventStream impl.

**Review by barkbarkimashark — COMMENTED** [review]
Since you're making changes to post_processing, I would add tests to verify the method call params are passed to the right functions:
tests/sentry/eventstream/test_eventstream.py
tests/sentry/eventstream/kafka/test_consumer.py
tests/sentry/eventstream/kafka/test_postprocessworker.py
tests/sentry/event_manager/test_event_manager.py

File: src/sentry/eventstream/kafka/postprocessworker.py (hunk 1 above)
Comment (barkbarkimashark) [reply src/sentry/eventstream/kafka/postprocessworker.py:96]: Is `SnubaEventStream` only used in dev stuff to avoid running kafka and zookeeper?

**Comment by MeredithAnya** [comment]
there's not too much context here so apologies if there is already information somewhere that I don't have, but are we ever going to consider having a separate task all together? I mean obv there is shared logic, but I'm assuming over time there are going to be very specific performance issue processing that errors don't need

**Review by lynnagara — CHANGES_REQUESTED** [review]
Rather than reusing the same Celery task for both use cases and then passing the queue to it, would it make sense to have a separate task for each error category instead (which can share parts of their implementations anyway)? It seems like there will some differences in post processing between performance and error groups.

File: src/sentry/tasks/post_process.py [hunk 2]
```diff
@@ -763,6 +763,15 @@ def plugin_post_process_group(plugin_slug, event, **kwargs):
     )
 
 
+def get_post_process_queue(group_category: GroupCategory) -> str:
+    if group_category == GroupCategory.ERROR:
+        return "post_process_errors"
+    elif group_category == GroupCategory.PERFORMANCE:
+        return "post_process_performance"
+    else:
+        return "post_process_errors"  # use error queue as default
```
Comment (lynnagara) [inline src/sentry/tasks/post_process.py:772]: Why is this needed? Shouldn't the category always be provided?

File: src/sentry/eventstream/kafka/postprocessworker.py (hunk 1 above)
Comment (lynnagara) [reply src/sentry/eventstream/kafka/postprocessworker.py:96]: Yes, it's primarily used in dev

**Comment by rjo100** [comment]
> Since you're making changes to post_processing, I would add tests to verify the method call params are passed to the right functions: tests/sentry/eventstream/test_eventstream.py tests/sentry/eventstream/kafka/test_consumer.py tests/sentry/eventstream/kafka/test_postprocessworker.py tests/sentry/event_manager/test_event_manager.py

Yeah was going to add testing to it. Probably should have been in draft still but wanted to get some feedback

File: src/sentry/tasks/post_process.py (hunk 2 above)
Comment (rjo100) [reply src/sentry/tasks/post_process.py:772]: not the cleanest code but generally would like to have a failsafe here. can remove the initial check for ERROR category for now

File: src/sentry/event_manager.py [hunk 3]
```diff
@@ -1082,12 +1083,14 @@ def _eventstream_insert_many(jobs):
             is_new = False
             is_regression = False
             is_new_group_environment = False
+            group_category = GroupCategory.ERROR
```
Comment (lynnagara) [inline src/sentry/event_manager.py:1086]: Is this absolutely needed? In other places there's the event type property that's used extensively to switch between the error vs transaction processing / post processing. This feels like extra data we are adding to every event for no reason that then gets passed through many steps.

File: src/sentry/tasks/post_process.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/tasks/post_process.py:772]: The signature of this method indicates it's argument can only be a GroupCategory, if that's not true then the typing is probably not correct.

Comment (rjo100) [reply src/sentry/tasks/post_process.py:772]: right but there are potentially more error categories in the future so I wanted to be defensive in case it gets overlooked or we want to spin it up quickly

File: src/sentry/event_manager.py (hunk 3 above)
Comment (wedamija) [reply src/sentry/event_manager.py:1086]: We don't want to rely on event level data to understand this, it's a group level property. We don't have any specific plans for this, but there's no reason that a different issue type couldn't be powered by transactions, if we found a valid use case.

**Review by wedamija — COMMENTED** [review]
I don't think this is actually inserting the `group_category` into the kafka message is it? 

Inserting/sending happens here: https://github.com/getsentry/sentry/blob/789d91b4cbadfb3ac65efffb85553789b1341466/src/sentry/eventstream/snuba.py#L163-L193

This is one reason that it would be simpler to add this to the `GroupState` rather than as a separate parameter, since it'd be automatically sent along, although I'm not sure what the serialization will look like.

File: src/sentry/event_manager.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/event_manager.py:1086]: Ok

File: src/sentry/eventstream/base.py [hunk 4]
```diff
@@ -71,20 +72,24 @@ def _dispatch_post_process_group_task(
         primary_hash: Optional[str],
         skip_consume: bool = False,
         group_states: Optional[GroupStates] = None,
+        group_category: Optional[GroupCategory] = GroupCategory.ERROR,
```
Comment (lynnagara) [inline src/sentry/eventstream/base.py:75]: Doesn't need to be optional since you have default value which is a GroupCategory.

File: src/sentry/event_manager.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/event_manager.py:1086]: But this serves no purpose today except as a tool to select a queue anyway. That doesn't seem like a group level thing as events without groups still go through that processing.

Also, isn't this being set anyway on events without a group here?

File: src/sentry/eventstream/base.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/eventstream/base.py:75]: Don't events without any groups still come through this function? I'm not sure that group category makes a ton of sense here when there is no group.

File: src/sentry/eventstream/base.py [hunk 5]
```diff
@@ -71,20 +72,24 @@ def _dispatch_post_process_group_task(
...
+                kwargs={
+                    "is_new": is_new,
+                    "is_regression": is_regression,
+                    "is_new_group_environment": is_new_group_environment,
+                    "primary_hash": primary_hash,
+                    "cache_key": cache_key,
+                    "group_id": group_id,
+                    "group_states": group_states,
+                },
+                queue=get_post_process_queue(group_category),
```
Comment (barkbarkimashark) [inline src/sentry/eventstream/base.py:92]: Small thing about apply_async, I think you gotta put the `queue` kwarg within `kwargs` since `post_process_group` is an instrumented_task and has the `queue` param pinned already. 

There seems to already be discussion on creating a separate task altogether for different categories, but this change will fix some of the failing tests.

**Comment by wedamija** [comment]
> Rather than reusing the same Celery task for both use cases and then passing the queue to it, would it make sense to have a separate task for each error category instead (which can share parts of their implementations anyway)? It seems like there will some differences in post processing between performance and error groups.

No, I'd prefer to keep the same task rather than requiring every issue type to define their own separate task. Just seems like boilerplate.

File: src/sentry/eventstream/base.py (hunk 4 above)
Comment (wedamija) [reply src/sentry/eventstream/base.py:75]: That's a reasonable point. I'm not sure if we should have a separate queue for issueless events, but maybe for now it does make sense to just rely on the event type

Comment (wedamija) [reply src/sentry/eventstream/base.py:75]: Either way though, we're going to have to pass something through here. By the time we get to https://github.com/getsentry/sentry/blob/28cbc543bbff5ca02972772b8c32c211d0fff199/src/sentry/eventstream/kafka/postprocessworker.py#L68 we already only have event ids, and I think it's preferable to not load things from the cache at this point?

File: src/sentry/eventstream/base.py (hunk 5 above)
Comment (rjo100) [reply src/sentry/eventstream/base.py:92]: I think celery overrides the `kwargs` argument to pass to the function and uses as `options` for the same effect

**Comment by lynnagara** [comment]
@wedamija My point is not that each issue type needs a queue, it's that the queue selection doesn't really seem to have much to do with issue types so calling it issue type seems kind of a misnomer. Every "error" and "transaction" message needs to get forwarded to post processing and must go to a queue regardless of whether than underlying event has one group attached to it, multiple groups attached to it or none. In the future, it could even be that one event has multiple groups of different issue types attached to it. If the queue depended on the issue type where would we even put that?

**Comment by wedamija** [comment]
> @wedamija My point is not that each issue type needs a queue, it's that the queue selection doesn't really seem to have much to do with issue types so calling it issue type seems kind of a misnomer. Every "error" and "transaction" message needs to get forwarded to post processing and must go to a queue regardless of whether than underlying event has one group attached to it, multiple groups attached to it or none. In the future, it could even be that one event has multiple groups of different issue types attached to it. If the queue depended on the issue type where would we even put that?

I don't mind too much if we want to rename these to error/transaction, it probably makes more sense. But I do just want to keep the one task, rather than require a new task per each supported data type.

**Comment by rjo100** [comment]
OK going to rename the queues to more appropriately reflect "error" vs "transaction" and move the queue info to the headers, as discussed during our team meeting

File: src/sentry/eventstream/snuba.py [hunk 6]
```diff
@@ -427,6 +427,9 @@ def insert(
         group_states: Optional[GroupStates] = None,
         **kwargs: Any,
     ) -> None:
+        message_type = "transaction" if self._is_transaction_event(event) else "error"
```
Comment (rjo100) [inline src/sentry/eventstream/snuba.py:430]: this might be adding a db call where we don't want one. need to discuss further

**Comment by rjo100** [comment]
still need to fix + add tests

File: src/sentry/eventstream/base.py [hunk 7]
```diff
@@ -43,6 +42,13 @@ class GroupState(TypedDict):
 GroupStates = Sequence[GroupState]
 
 
+def get_post_process_queue(message_type: str) -> str:
+    if message_type == "transaction":
+        return "post_process_transactions"
```
Comment (udameli) [inline src/sentry/eventstream/base.py:47]: Has `post_process_transactions` queue been created? I think right now there's `post_process_performance`

File: src/sentry/eventstream/kafka/backend.py [hunk 8]
```diff
@@ -125,6 +125,8 @@ def insert(
                 {"project_id": event.project_id, "message_type": message_type},
             )
 
+        kwargs["message_type"] = message_type
+
```
Comment (wedamija) [inline src/sentry/eventstream/kafka/backend.py:129]: Where do we pass this to Kafka?

**Review by wedamija — COMMENTED** [review]
You'll need to include this message type in the headers here, and probably the body as well just for consistency. The process in production here looks like:

 - Eventstream.insert produces a message to kafka
 - Post process forwarder receives the message, parses the headers and then fires the post process group task

In dev it looks like
 - Eventstream insert calls snuba directly with the data, then fires off post process group task in process

File: src/sentry/eventstream/base.py (hunk 7 above)
Comment (rjo100) [reply src/sentry/eventstream/base.py:47]: just fixed it in code. was an oversight that wasn't merged. do we need to circle back to rename it on an infrastructure level?

File: src/sentry/eventstream/kafka/backend.py (hunk 8 above)
Comment (rjo100) [reply src/sentry/eventstream/kafka/backend.py:129]: might have accidentally removed it. one sec

Comment (rjo100) [reply src/sentry/eventstream/kafka/backend.py:129]: in the insert call below? or am I mistaken?

Comment (wedamija) [reply src/sentry/eventstream/kafka/backend.py:131]: This is the insert call we make here: https://github.com/getsentry/sentry/blob/6b4bbc70d51626614b2909656750b322b5b1d5a2/src/sentry/eventstream/snuba.py#L112-L123

The kwargs aren't directly used for anything, so `message_type` is just thrown into the void. It is used in the Snuba backend, but since we're passing a message to a consumer, `message_type` will need to be included as part of the headers, since that's what the consumer reads.

File: src/sentry/eventstream/base.py [hunk 9]
```diff
@@ -71,20 +78,24 @@ def _dispatch_post_process_group_task(
         primary_hash: Optional[str],
         skip_consume: bool = False,
         group_states: Optional[GroupStates] = None,
+        message_type: str = "error",
```
Comment (lynnagara) [inline src/sentry/eventstream/base.py:81]: I don't think message_type "error" is a reasonable default here, please ensure a value is passed instead.

File: src/sentry/eventstream/base.py (hunk 7 above)
Comment (lynnagara) [reply src/sentry/eventstream/base.py:47]: I think this function adds an unnecessary level of indirection. Why mapping to a message type string first before the queue - would it make more sense to just pass the queue name? Especially since the message type is not really right and just seems to default to "error" anyway in most places.

File: src/sentry/eventstream/base.py [hunk 10]
```diff
@@ -97,6 +108,7 @@ def insert(
         received_timestamp: float,
         skip_consume: bool = False,
         group_states: Optional[GroupStates] = None,
+        **kwargs: Any,
```
Comment (lynnagara) [inline src/sentry/eventstream/base.py:111]: I don't think you should do this. Why can't you use: `Event.get_event_type() == "transaction"` as all the other checks in the post process forwarder do?

**Review by lynnagara — CHANGES_REQUESTED** [review]
Nearly all processing uses `Event.get_event_type() == "transaction"` to differentiate between errors and transactions case. While it may not be ideal, I think it's better to stick to this than try to pass another message type parameter all the way through a bunch of functions for the sole purpose of selecting a queue.

As for separate tasks, I think this is likely the long term direction. The eventstream only supports errors and transaction types and it would be a pretty herculean effort to get another event type in there and I don't think we would do that. All of our new event types never piggyback the eventstream abstraction anymore. More likely it will actually split into separate errors and transactions variants anyway (perhaps something along the lines of https://github.com/getsentry/sentry/pull/37376) and thus a separate task may be desirable. But I won't block on this if you want to keep it the same right now.

File: src/sentry/eventstream/base.py (hunk 9 above)
Comment (wedamija) [reply src/sentry/eventstream/base.py:81]: > Nearly all processing uses `Event.get_event_type() == "transaction"` to differentiate between errors and transactions case. While it may not be ideal, I think it's better to stick to this than try to pass another message type parameter all the way through a bunch of functions for the sole purpose of selecting a queue.
> 
> As for separate tasks, I think this is likely the long term direction. The eventstream only supports errors and transaction types and it would be a pretty herculean effort to get another event type in there and I don't think we would do that. All of our new event types never piggyback the eventstream abstraction anymore. More likely it will actually split into separate errors and transactions variants anyway (perhaps something along the lines of #37376) and thus a separate task may be desirable. But I won't block on this if you want to keep it the same right now.

I don't mind what we pass too much, but we have to pass something through the headers to the ppf so that it can differentiate and decide which queue to pass to. I'm happy for that to just be the event type. 

For separate tasks, we might not be using the ppf here but we are likely sticking with the same task that will be fired from somewhere else. We'll see how this evolves as we work with the profiling team though.

**Comment by lynnagara** [comment]
Should this be closed since this looks to be reimplemented in https://github.com/getsentry/sentry/pull/39946?

**Comment by rjo100** [comment]
yeah was going to. thx
